docs/DafnyRef/Refinement.md

```diff
@@ -92,32 +92,205 @@ TODO
+  ghost const ToGhost : int
+  const NewConst : int
+}
+
+
+```
+
+Formally, a child `const` declaration may refine a `const` declaration
+from a parent module if
 
 * the parent has no initialization,
 * the child has the same type as the parent, and
 * one or both of the following holds:
    * the child has an initializing expression
-   * the child is declared `ghost` and the parent is not `ghost`, or vice versa
+   * the child is declared `ghost` and the parent is not `ghost`.
 
-To indicate it is a refining declaration, a refining `const` declaration
-contains an ellipsis in this syntax:
-```grammar
-"const" { Attribute } CIdentType "..." [ ":=" Expression ]
-```
+A refining module can also introduce new `const` declarations that do
+not exist in the refinement parent.
 
 ## 21.5. Method declarations
 
-TODO
+Method declarations can be refined as in the following example.
+
+```
+module A {
+  method ToImplement(x: int) returns (r: int)
+    ensures r > x
+
+  method ToStrengthen(x: int) returns (r: int)
+
+  method ToDeterminize(x: int) returns (r: int)
+    ensures r >= x
+  {
+    var y :| y >= x;
+    return y;
+  }
+
+  method ToSuperimpose(x: int) returns (r: int)
+  {
+    if (x < 0) {
+      return -x;
+    } else {
+      return x;
+    }
+  }
+
+}
+
+module B refines A {
+  method ToImplement(x: int) returns (r: int)
+  {
+    return x + 2;
+  }
+
+  method ToStrengthen...
+    ensures r == x*2
+  {
+    return x*2;
+  }
+
+  method ToDeterminize(x: int) returns (r: int)
+  {
+    return x;
+  }
+
+  method ToSuperimpose(x: int) returns (r: int)
+  {
+    if (x < 0) {
+      print "inverting";
+    } else {
+      print "not modifying";
+    }
+  }
+}
+```
+
+Formally, a child `method` definition may refine a parent `method`
+declaration or definition to perform one or more of the following
+operations:
+
+* provide a body missing in the parent (as in `ToDefine`),
+* strengthen the postcondition of the parent method by adding one or more
+  `ensures` clauses (as in `ToStrengthen`),
+* provide a more deterministic version of a non-deterministic parent
+  body (as in `ToDeterminize`), or
+* superimpose the body of the parent method with additional statements
+  (as in `ToSuperimpose`).
+
+The type signature of a refining method must be the same as that of the
+parent method it refines. This can be ensured by providing an explicit
+type signature equivalent to that of the parent (with renaming of
+parameters allowed) or by using an ellipsis (`...`) to indicate copying
+of the parent type signature. The following example illustrates that
+latter case.
+
+To introduce additional statements, the child method can include
+ellipses within the body to stand in for portions of code from the
+parent body. Dafny then attempts to merge the body of the child with the
+body of the parent by filling in the ellipses. In the `ToSuperimpose`
+example, the explicit `...` at the beginning will expand to the variable
```

Comment:
Where is the explicit "..." at the beginning?

Reply:
Fixed! I had a local file I was testing that wasn't fully reflected in the text.